### PR TITLE
Build nuget pkg for ROCm

### PR DESCRIPTION
**Pull request by cloudhan**
Add nuget pkg building and publishing for ROCm EP

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml [hunk 1]
```diff
@@ -775,6 +813,222 @@ stages:
...
+        BuildId: ${{ parameters.BuildId }}
+
+    - task: PowerShell@2
+      displayName: 'Reconstruct Build Directory'
+      inputs:
+        targetType: inline
+        script: |
+          Get-ChildItem $(Build.BinariesDirectory)\nuget-artifact -Filter *.tgz | % {
+            # *.tar will be created after *.tgz is extracted
+            $cmd = "7z.exe x $($_.FullName) -y -o$(Build.BinariesDirectory)\nuget-artifact"
```
Comment (yuslepukhin) [inline tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml:852]: >"7z.exe [](http://example.com/codeflow?start=19&length=7)

Will this run on Linux?

File: tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml [hunk 2]
```diff
@@ -792,6 +1046,15 @@ stages:
     SpecificArtifact: ${{ parameters.specificArtifact }}
     BuildId: ${{ parameters.BuildId }}
 
+# - template: nuget/templates/test_linux.yml
```
Comment (yuslepukhin) [inline tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml:1049]: > - template: [](http://example.com/codeflow?start=1&length=12)

Should this be deleted?

File: tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml [hunk 3]
```diff
@@ -220,6 +220,51 @@ stages:
...
+      clean: all
+    timeoutInMinutes: 120
+    pool: onnxruntime-Ubuntu2004-AMD-CPU
+    variables:
+      RocmVersion: '5.5'
+    steps:
+    - template: templates/set-version-number-variables-step.yml
+    - template: templates/get-docker-image-steps.yml
+      parameters:
+        Dockerfile: tools/ci_build/github/pai/rocm-ci-pipeline-env.Dockerfile
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/c-api-noopenmp-packaging-pipelines.yml:238]: But this is an Ubuntu file? All other build jobs use CentOS.

**Comment by cloudhan** [comment]
@snnn @mszhanyi  @yuslepukhin I think this one is finally ready.

**Comment by snnn** [comment]
Please also add https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/LICENSE to https://github.com/microsoft/onnxruntime/blob/main/ThirdPartyNotices.txt . It can be in a new PR.

**Comment by mszhanyi** [comment]
Could you try and(eq(variables['Agent.OS'], 'Linux'), not(contains(variables['Agent.JobName'], 'ROCM'))) to fix the warning.
https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/azure-pipelines/templates/clean-agent-build-directory-step.yml

**Review by mszhanyi — APPROVED** [review]

File: tools/ci_build/github/azure-pipelines/templates/publish-nuget.yml [hunk 4]
```diff
@@ -11,6 +11,8 @@ stages:
   - NuGet_Test_Linux_CPU
   - NuGet_Test_Win_GPU
   - NuGet_Test_Linux_GPU
+  - NuGet_Packaging_ROCm # TODO use NuGet_Test_Linux_ROCm once we have agent pool for testing
+  # - NuGet_Test_Linux_ROCm
```
[the remark this answers is not in the document]
Comment (cloudhan) [reply tools/ci_build/github/azure-pipelines/templates/publish-nuget.yml]: Oops

**Review by mszhanyi — APPROVED** [review]

**Pull request merged**